From 58eab50cc1a429e5808f6b3e6f71fe11da6c91af Mon Sep 17 00:00:00 2001
From: Kent Overstreet <koverstreet@google.com>
Date: Sun, 1 Jun 2014 17:21:13 -0700
Subject: [PATCH 064/100] block: Add bio_for_each_segment_all()

__bio_for_each_segment() iterates bvecs from the specified index
instead of bio->bv_idx.  Currently, the only usage is to walk all the
bvecs after the bio has been advanced by specifying 0 index.

For immutable bvecs, we need to split these apart;
bio_for_each_segment() is going to have a different implementation.
This will also help document the intent of code that's using it -
bio_for_each_segment_all() is only legal to use for code that owns the
bio.
[patch reduced to just include/linux/bio change]

Change-Id: Ia2f23d32db6149223beda41bfe7956bb73e1c1dc

Signed-off-by: artas182x <artas182x@gmail.com>
---
 include/linux/bio.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/include/linux/bio.h b/include/linux/bio.h
index ce33e68..da350c4 100644
--- a/include/linux/bio.h
+++ b/include/linux/bio.h
@@ -143,6 +143,15 @@ static inline int bio_has_allocated_vec(struct bio *bio)
 	     i < (bio)->bi_vcnt;					\
 	     bvl++, i++)
 
+/*
+ * drivers should _never_ use the all version - the bio may have been split
+ * before it got to the driver and the driver won't own all of it
+ */
+#define bio_for_each_segment_all(bvl, bio, i)				\
+	for (i = 0;							\
+	     bvl = bio_iovec_idx((bio), (i)), i < (bio)->bi_vcnt;	\
+	     i++)
+
 #define bio_for_each_segment(bvl, bio, i)				\
 	__bio_for_each_segment(bvl, bio, i, (bio)->bi_idx)
 
-- 
1.9.1

